packages/@tailwindcss-upgrade/src/codemods/format-nodes.ts

```diff
@@ -1,33 +1,73 @@
-import { parse, type ChildNode, type Plugin, type Root } from 'postcss'
+import postcss, { type ChildNode, type Plugin, type Root } from 'postcss'
 import { format } from 'prettier'
-import { walk, WalkAction } from '../utils/walk'
+import { walk } from '../utils/walk'
 
 // Prettier is used to generate cleaner output, but it's only used on the nodes
 // that were marked as `pretty` during the migration.
 export function formatNodes(): Plugin {
   async function migrate(root: Root) {
     // Find the nodes to format
     let nodesToFormat: ChildNode[] = []
-    walk(root, (child) => {
-      if (child.raws.tailwind_pretty) {
+    walk(root, (child, _idx, parent) => {
+      // Always print semicolons after at-rules
+      if (child.type === 'atrule') {
+        child.raws.semicolon = true
+      }
+
+      if (child.type === 'atrule' && child.name === 'tw-bucket') {
         nodesToFormat.push(child)
-        return WalkAction.Skip
+      } else if (child.raws.tailwind_pretty) {
+        // @ts-expect-error We might not have a parent
+        child.parent ??= parent
+        nodesToFormat.unshift(child)
       }
     })
 
+    let output: string[] = []
+
     // Format the nodes
-    await Promise.all(
-      nodesToFormat.map(async (node) => {
-        node.replaceWith(
-          parse(
-            await format(node.toString(), {
-              parser: 'css',
-              semi: true,
-              singleQuote: true,
-            }),
-          ),
-        )
-      }),
+    for (let node of nodesToFormat) {
+      let contents = (() => {
+        if (node.type === 'atrule' && node.name === 'tw-bucket') {
+          // Remove the `@tw-bucket` wrapping, and use the contents directly.
+          return node
+            .toString()
+            .trim()
+            .replace(/@tw-bucket(.*?){([\s\S]*)}/, '$2')
+        }
+
+        return node.toString()
+      })()
+
+      // Do not format the user bucket to ensure we keep the user's formatting
+      // intact.
+      if (node.type === 'atrule' && node.name === 'tw-bucket' && node.params === 'user') {
+        output.push(contents)
+        continue
+      }
+
+      // Format buckets
+      if (node.type === 'atrule' && node.name === 'tw-bucket') {
+        output.push(await format(contents, { parser: 'css', semi: true, singleQuote: true }))
```

Comment:
Maybe we can move out this part into a const object so we ensure we don't diverge the prettier config between the two branches (and avoid creating the same object over and over):

```suggestion
        output.push(await format(contents, FORMATTING_OPTIONS))
```